_source/_posts/2022-09-26-java-jetty-oauth.md

```diff
@@ -0,0 +1,1017 @@
+```bash
+http :8080/hikes
+```
+```
+HTTP/1.1 200 OK
+...
+
+Wonderland Trail
+South Maroon Peak
+Tour du Mont Blanc
+Teton Crest Trail
+Everest Base Camp via Cho La Pass
+Kesugi Ridge
+```
+
+POST a new hike:
+
+```bash
+http -f POST :8080/hikes hike="Pear Lake"
+```
+
+```
+HTTP/1.1 200 OK
+...
+
+Wonderland Trail
+South Maroon Peak
+Tour du Mont Blanc
+Teton Crest Trail
+Everest Base Camp via Cho La Pass
+Kesugi Ridge
+Pear Lake
+
+```
+
+DELETE a hike:
+
+```bash
+http DELETE :8080/hikes hike=="South Maroon Peak"
+```
+
+```
+HTTP/1.1 200 OK
+Content-Length: 110
+Date: Mon, 05 Sep 2022 21:01:34 GMT
+Server: Jetty(11.0.11)
+
+Wonderland Trail
+Tour du Mont Blanc
+Teton Crest Trail
+Everest Base Camp via Cho La Pass
+Kesugi Ridge
+Pear Lake
+```
+
+Now try to delete a hike that doesn't exist or send an empty value:
+
+```bash
+http DELETE :8080/hikes
+```
+
+```
+HTTP/1.1 400 Bad Request
+...
+
+Param 'hike' cannot be null.
+```
+
+That's how you can create a web service using Maven and Jetty. To deploy this, you would typically build the packaged WAR file using `mvn package` and deploy the WAR to your Jetty server.
+
+> **NOTE:** this is a very naive implementation of a REST service. It uses an in-memory `ArrayList` as a data source, which is not synchronized (and thus would run into threading problems in a real web servlet). For anything beyond the scope of this tutorial, you'd need to implement a database backend of some kind. For help on how to do this, see the example blog posts listed at the end of the tutorial. You would typically also add a PUT endpoint and assign each item an ID to use as an index so data can be updated, but that is beyond the scope of this tutorial.
+
+## Build a web service with Gradle and Jetty
+
+Next, you're going to look at how you can accomplish the same thing you did above, but this time with Gradle instead of Maven. Why? Because some people are allergic to XML and love DSLs instead.
+
+The Gradle-based app is located in the `gradle-jetty` subdirectory.
+
+[Gradle](https://docs.gradle.org/current/userguide/userguide.html) is a Groovy-based build system that uses a custom DSL (domain-specific language) for its build files. Because the build script files are essentially Groovy scripts, the build system is incredibly powerful, thus, very flexible.
+
+To run a Jetty server using Gradle, you'll use [the Gretty plugin](https://github.com/gretty-gradle-plugin/gretty).
+
+From the Gretty docs:
+
+> Gretty is a feature-rich Gradle plugin for running web apps on embedded servlet containers. It supports Jetty versions 7, 8, and 9, Tomcat versions 7 and 8, multiple web apps, and many more. It wraps servlet  container functions as convenient Gradle tasks and configuration DSL
+
+This project uses Gretty plugin version 4.0.3, which uses Jetty version 11.0.11. Since version 4, Gretty requires Jakarta over Javax. You can use Gretty 3 if you want to support non-Jakarta apps (this is explained on the Gretty GitHub page above).
```

Comment:
```suggestion
This project uses Gretty plugin version 4.0.3, which uses Jetty version 11.0.11. Since version 4, Gretty requires `jakarta` imports over `javax`. You can use Gretty 3 if you want to support non-Jakarta apps (this is explained on the Gretty GitHub page above).
```